src/video/aml.c

```diff
@@ -145,31 +163,43 @@ void aml_cleanup() {
   }
 
   codec_close(&codecParam);
+  free(frame_buffer);
 }
 
 int aml_submit_decode_unit(PDECODE_UNIT decodeUnit) {
+
+  if (decodeUnit->fullLength > DECODER_BUFFER_SIZE) {
+    fprintf(stderr, "Video decode buffer too small, %i > %i\n", decodeUnit->fullLength, DECODER_BUFFER_SIZE);
+    return DR_OK;
+  }
+
+  int result = DR_OK, length = 0, errCounter = 0, api;
   PLENTRY entry = decodeUnit->bufferList;
-  while (entry != NULL) {
-    char* data = entry->data;
-    int length = entry->length;
-
-    while (length > 0) {
-      int written = codec_write(&codecParam, data, length);
-      if (written > 0) {
-        data += written;
-        length -= written;
-      } else if (errno == EAGAIN) {
-        usleep(500);
-      } else {
-        fprintf(stderr, "codec_write() failed: %d\n", errno);
+  do {
+    memcpy(frame_buffer+length, entry->data, entry->length);
+    length += entry->length;
+    entry = entry->next;
+  } while (entry != NULL);
+
+  codec_checkin_pts(&codecParam, decodeUnit->presentationTimeMs);
+  do {
+    api = codec_write(&codecParam, frame_buffer, length);
```

Comment:
I think this can (theoretically) perform a partial write, so we need to handle the case where the amount written is less than the total, like the old code did.